### PR TITLE
refactor(storage): replace PbHummockVersion with new HummockVersion struct

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

In this PR, we introduce a new separate struct for `HummockVersion`, `HummockVersionDelta` and `HummockVersionCheckpoint`. This is a refactor PR, and currently the fields in these structs are the same as the counterpart pb structs. This is for the meta data change introduced in https://github.com/risingwavelabs/risingwave/issues/14041. The backward compatibility of meta data will be maintained in the `from_protobuf` method of these structs. In this PR we have limited all usages of their pb structs in rpc and metadata persistence. All other usages will be using the new struct.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by Little-Wallace — COMMENTED** [review]
Since `PbHummockVersionDelta` is only used to persist, is it necessary to replace it with a new struct ?

**Comment by wenym1** [comment]
> Since `PbHummockVersionDelta` is only used to persist, is it necessary to replace it with a new struct ?

The initial motivation for this PR is for the meta data change introduced in https://github.com/risingwavelabs/risingwave/issues/14041. In the issue we will change from using a single global `max_committed_epoch` to per-table `max_committed_epoch`. Since `HummockVersion` and `HummockVersionDelta` are persisted, this will be a breaking change and we need to maintain backward compatibility. So in this PR we implement a separate struct for these two metadata so that when we introduce the breaking change, we can maintain backward compatibility in the `from_protobuf` method and hide the difference between the two versions of metadata.

File: src/meta/src/hummock/metrics_utils.rs [hunk 1]
```diff
@@ -45,6 +45,7 @@ pub fn trigger_version_stat(
     metrics
         .max_committed_epoch
         .set(current_version.max_committed_epoch as i64);
+    // TODO: this op may be costly
```
Comment (zwang28) [inline src/meta/src/hummock/metrics_utils.rs:48]: Indeed we've seen version of tens of megabytes. Will any fix include in this PR?

**Review by zwang28 — APPROVED** [review]
rest LGTM.

File: src/meta/src/hummock/metrics_utils.rs (hunk 1 above)
Comment (wenym1) [reply src/meta/src/hummock/metrics_utils.rs]: I have changed to implement an `estimated_encoded_len` by ourselves.

File: src/storage/hummock_sdk/src/version.rs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+            safe_epoch: self.safe_epoch,
+            table_watermarks: self
+                .table_watermarks
+                .iter()
+                .map(|(table_id, watermark)| (table_id.table_id, watermark.to_protobuf()))
+                .collect(),
+        }
+    }
+
+    pub fn estimated_encode_len(&self) -> usize {
```
Comment (Li0k) [inline src/storage/hummock_sdk/src/version.rs:92]: Does the estimate of size need to correspond to all the member fields?

**Review by Li0k — APPROVED** [review]
Rest LGTM

File: src/storage/hummock_sdk/src/version.rs (hunk 2 above)
Comment (wenym1) [reply src/storage/hummock_sdk/src/version.rs:92]: This is just an estimated encode len. For simplicity I only include the encode len of fields with variable length (e.g. list and map).

File: src/storage/hummock_sdk/src/version.rs [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+impl HummockVersionDelta {
+    pub fn from_persisted_protobuf(delta: &PbHummockVersionDelta) -> Self {
+        Self::from_protobuf_inner(delta)
+    }
+
+    pub fn from_rpc_protobuf(delta: &PbHummockVersionDelta) -> Self {
+        Self::from_protobuf_inner(delta)
+    }
```
Comment (hzxa21) [inline src/storage/hummock_sdk/src/version.rs:134]: Need some docs to explain why we need to distinguish rpc protobuf and persisted protobuf.

**Review by hzxa21 — APPROVED** [review]
Rest LGTM!

**Pull request merged**